torch/csrc/autograd/variable.h

```diff
@@ -531,6 +531,11 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
   c10::optional<ViewInfo> backward_info_;
   c10::optional<ViewInfo> forward_info_;
 
+  // Optimization to reduce the number of ViewInfo we create.
+  // In the (very common) case where backward_info_ == forward_info_
+  // we only populate backward_info_ and set shared_view_info_ = true
+  bool shared_view_info_;
```

Comment:
What I wouldn't give for an algebraic data type here...

```
data ViewMeta
  = NoViewInfo
  | SharedViewInfo ViewInfo
  | DistinctViewInfo ViewInfo ViewInfo
```

Reply:
The comment here would be clearer if you explicitly spelled out the invariants expected to be respected at all times on ViewInfo.